public_html/wp-content/mu-plugins/blocks/assets/src/shared/featured-image/index.js

```diff
@@ -0,0 +1,183 @@
+	 * @param props.alt Alt text for image
+	 */
+	constructor( props ) {
+		super( props );
+	}
+
+	componentWillMount() {
+		let availableSizes;
+		const { wpMediaDetails } = this.props;
+		if ( wpMediaDetails ) {
+			availableSizes = FeaturedImage.parseWPMediaDetails( wpMediaDetails );
+		} else {
+			availableSizes = this.props.availableSizes;
+		}
+		this.parseAvailableSizes( availableSizes );
+	}
+
+	/**
+	 * Converts from WP API featured_media object to array of different sizes.
+	 *
+	 * @param wpMediaDetails
+	 */
+	static parseWPMediaDetails( wpMediaDetails ) {
+		const availableSizes = [];
+		for ( const size of Object.keys( wpMediaDetails ) ) {
+			if ( ! wpMediaDetails.hasOwnProperty( size  ) ) {
+				continue;
+			}
+
+			availableSizes.push( wpMediaDetails[ size ] );
+		}
+		return availableSizes;
+	}
+
+	/**
+	 * Pre-calculate and store aspect ratio of all images. This will help us in picking the best image available for the size that we have.
+	 */
+	parseAvailableSizes( availableSizes ) {
+		let largestSizeValue = 0;
+		let largestSize = {};
+		// Lets pre-calculate aspect ratio of all images. Also cache the largest area image.
+		const sizes = availableSizes.map( ( size ) => {
+			return {
+				aspectRatio: size.width / size.height,
+				area: size.width * size.height,
+				...size
+			}
+		} );
+
+		// Lets sort the available images based on area and then aspect ratio.
+		sizes.sort( ( size1, size2 ) => {
+				if ( size1.area < size2.area ) {
+					return -1;
+				} else if ( size1.area === size2.area ) {
+					return size1.aspectRatio < size2.aspectRatio ? -1 : 1;
+				} else {
+					return 1;
+				}
+			}
+		);
+
+		this.setState(
+			{
+				sizes,
+				...this.state
+			}
+		);
+	}
+
+
+	/**
+	 * Returns URL of appropriate image. We find appropriate by looking for an image with similar aspect ratio, and similar or greater area.
+	 * If there is no image with similar aspect ratio or area, we will return strictly bigger image then currently requested.
+	 * If there is no bigger image then currently request one, we will return the max size image URL.
+	 * Note: Core Image block allows user to select image size by name, but we cannot really do that because we will be controlling multiple images at once, and one or more images may not have all the available sizes. So an automated solution is preferred here.
+	 *
+	 * @param height
+	 * @param width
+	 */
+	getSizedUrl( height, width ) {
+		const { sizes } = this.state;
+		if ( ! sizes ) {
+			return "";
+		}
+
+		// Lets loop through all possible images, see if we have an image that matches requested aspect ration. if we have, we will return it. Also keep track of image which has the most closest aspect ration.
+		let smallestImage;
+		let selectedImage;
+		const requiredAspectRatio = width / height;
+
+		for ( const size of sizes ) {
+
+			// Reject all the images with less height or width then 20px.
+			if (size.height < height - 20 || size.width < width - 20) {
```

Comment:
Will this be sufficient to catch all images that aren't the right size? What if the height is within 20px but the width is way less?

Reply:
So if width is less then 20px, then the image will be rejected.

Reply:
This code was removed. Resolving.